File: backend/app/api/routes/llm.py
"""
LLM API routes.

Provides endpoints for getting AI-generated explanations of vulnerabilities
and security concepts.
"""

from fastapi import APIRouter, HTTPException, Query

from app.core.logging import get_api_logger
from app.services.llm import (
    LLMService,
    get_llm_service,
    ExplanationRequest,
    ExplanationResponse,
    ExplanationType,
)
from app.services.llm.models import LLMHealthStatus

logger = get_api_logger()

router = APIRouter(prefix="/llm", tags=["LLM"])


@router.post(
    "/explain",
    response_model=ExplanationResponse,
    summary="Get an explanation",
    description="Generate an AI-powered explanation for a security topic.",
)
async def get_explanation(
    request: ExplanationRequest,
    skip_cache: bool = Query(
        default=False,
        description="Skip cache lookup (force fresh generation)"
    ),
) -> ExplanationResponse:
    """
    Generate an explanation for a vulnerability or security concept.

    The service uses a fallback chain:
    1. Local Ollama (if available)
    2. Hosted API (if configured)
    3. Static knowledge base (always available)

    Results are cached to improve performance.
    """
    logger.info(
        "Explanation request received",
        extra={
            "topic": request.topic,
            "type": request.explanation_type.value,
            "difficulty": request.difficulty_level,
        }
    )

    try:
        service = get_llm_service()
        response = await service.get_explanation(request, skip_cache=skip_cache)

        logger.info(
            "Explanation generated",
            extra={
                "topic": request.topic,
                "provider": response.provider.value,
                "cached": response.cached,
            }
        )

        return response

    except Exception as e:
        logger.error(f"Error generating explanation: {e}")
        raise HTTPException(
            status_code=500,
            detail="Failed to generate explanation. Please try again later."
        )


@router.get(
    "/explain/vulnerability/{vuln_type}",
    response_model=ExplanationResponse,
    summary="Explain a vulnerability type",
    description="Get an explanation for a specific vulnerability type.",
)
async def explain_vulnerability(
    vuln_type: str,
    difficulty: str = Query(
        default="beginner",
        pattern="^(beginner|intermediate|advanced)$",
        description="Explanation difficulty level"
    ),
    context: str | None = Query(
        default=None,
        max_length=500,
        description="Additional context about the vulnerability"
    ),
) -> ExplanationResponse:
    """
    Shortcut endpoint to explain a vulnerability type.

    Provides a convenient way to get vulnerability explanations
    without constructing a full request body.
    """
    request = ExplanationRequest(
        explanation_type=ExplanationType.VULNERABILITY,
        topic=vuln_type,
        context=context,
        difficulty_level=difficulty,
    )

    service = get_llm_service()
    return await service.get_explanation(request)


@router.get(
    "/explain/remediation/{vuln_type}",
    response_model=ExplanationResponse,
    summary="Explain remediation steps",
    description="Get remediation steps for a specific vulnerability type.",
)
async def explain_remediation(
    vuln_type: str,
    difficulty: str = Query(
        default="beginner",
        pattern="^(beginner|intermediate|advanced)$",
        description="Explanation difficulty level"
    ),
    context: str | None = Query(
        default=None,
        max_length=500,
        description="Additional context about the environment"
    ),
) -> ExplanationResponse:
    """
    Shortcut endpoint to get remediation instructions.

    Provides step-by-step instructions for fixing a vulnerability.
    """
    request = ExplanationRequest(
        explanation_type=ExplanationType.REMEDIATION,
        topic=vuln_type,
        context=context,
        difficulty_level=difficulty,
    )

    service = get_llm_service()
    return await service.get_explanation(request)


@router.get(
    "/explain/concept/{concept}",
    response_model=ExplanationResponse,
    summary="Explain a security concept",
    description="Get an explanation for a security concept.",
)
async def explain_concept(
    concept: str,
    difficulty: str = Query(
        default="beginner",
        pattern="^(beginner|intermediate|advanced)$",
        description="Explanation difficulty level"
    ),
) -> ExplanationResponse:
    """
    Shortcut endpoint to explain a security concept.

    Provides educational explanations of cybersecurity concepts.
    """
    request = ExplanationRequest(
        explanation_type=ExplanationType.CONCEPT,
        topic=concept,
        difficulty_level=difficulty,
    )

    service = get_llm_service()
    return await service.get_explanation(request)


@router.get(
    "/health",
    response_model=LLMHealthStatus,
    summary="Check LLM service health",
    description="Check the availability of LLM providers.",
)
async def check_health() -> LLMHealthStatus:
    """
    Check the health status of LLM providers.

    Returns information about:
    - Ollama availability
    - Hosted API availability
    - Currently active provider
    - Cache size
    """
    service = get_llm_service()
    return await service.check_health()


@router.get(
    "/cache/stats",
    summary="Get cache statistics",
    description="Get statistics about the explanation cache.",
)
async def get_cache_stats() -> dict:
    """
    Get cache statistics.

    Returns:
    - Cache hits and misses
    - Hit rate
    - Current cache size
    - Number of evictions
    """
    service = get_llm_service()
    return service.get_cache_stats()


@router.post(
    "/cache/clear",
    summary="Clear the cache",
    description="Clear all cached explanations.",
)
async def clear_cache() -> dict:
    """
    Clear the explanation cache.

    Use this if you need to force fresh explanations
    or if cached content is outdated.
    """
    logger.info("Cache clear requested")

    service = get_llm_service()
    count = service.clear_cache()

    return {
        "message": "Cache cleared successfully",
        "entries_cleared": count,
    }

File: frontend/src/types/llm.ts
/**
 * LLM-related types.
 *
 * These types define the data structures for LLM explanations
 * and AI-powered educational content.
 */

import type { DifficultyLevel } from './scenario';

/**
 * Types of explanations available from the LLM service.
 */
export type ExplanationType =
  | 'vulnerability'
  | 'remediation'
  | 'concept'
  | 'service'
  | 'risk';

/**
 * LLM provider types.
 */
export type LLMProvider = 'ollama' | 'hosted' | 'static';

/**
 * Request for an LLM explanation.
 */
export interface ExplanationRequest {
  /** Type of explanation requested */
  explanation_type: ExplanationType;
  /** The topic to explain */
  topic: string;
  /** Additional context for better explanation */
  context?: string;
  /** Target difficulty level */
  difficulty_level: DifficultyLevel;
}

/**
 * Response containing an LLM-generated explanation.
 */
export interface ExplanationResponse {
  /** The generated explanation text */
  explanation: string;
  /** Which LLM provider generated this explanation */
  provider: LLMProvider;
  /** The topic that was explained */
  topic: string;
  /** Whether this response was served from cache */
  cached: boolean;
  /** The difficulty level of the explanation */
  difficulty_level: DifficultyLevel;
  /** Suggested related topics to explore */
  related_topics: string[];
}

/**
 * Health status of LLM providers.
 */
export interface LLMHealthStatus {
  /** Whether local Ollama instance is available */
  ollama_available: boolean;
  /** Whether hosted LLM API is available */
  hosted_available: boolean;
  /** Currently active LLM provider */
  active_provider: LLMProvider;
  /** Number of explanations in cache */
  cache_size: number;
}

/**
 * Display-friendly provider names.
 */
export const PROVIDER_NAMES: Record<LLMProvider, string> = {
  ollama: 'Local AI (Ollama)',
  hosted: 'Cloud AI',
  static: 'Knowledge Base',
};

/**
 * Display-friendly difficulty descriptions.
 */
export const DIFFICULTY_DESCRIPTIONS: Record<DifficultyLevel, string> = {
  beginner: 'Simple explanations for newcomers',
  intermediate: 'Technical details for learners',
  advanced: 'In-depth analysis for professionals',
  expert: 'Expert-level technical deep dives',
};
